### PR TITLE
Add PyPubSubIslandEventQueue

### LINKED ISSUE #2233: Implement IslandEventQueue
## Description
The Island needs an internal pub/sub event queue to drive data processing and other server-side activities. Create and implement `IIslandEventQueue` (interface shown below) using `pypubsub`.

```python
from abc import ABC, abstractmethod
from enum import Enum, auto
from typing import Any, Callable, Optional


class IslandEventTopics(Enum):
    AGENT_CONNECTED = auto()
    CLEAR_SIMULATION_DATA = auto()
    RESET_AGENT_CONFIGURATION = auto()

class IIslandEventQueue(ABC):
    @abstractmethod
    def publish(self, topic: IslandEventTopics, event: Optional[Any] = None):
        pass

    @abstractmethod
    def subscribe(self, topic: IslandEventTopics, subscriber: Callable[[...], None]):
        pass
```

## Tasks

- [ ] Create `IIslandEventQueue` interface (0.25d) - @shreyamalviya 
    - [x] Create `IslandEventTopics` Enum
    - [x] Add docstrings
- [ ] Implement `PyPubSubIslandEventQueue` (0.25d) - @shreyamalviya 
- [x] Add `PyPubSubEventQueue` to DIContainer (0d) - @shreyamalviya 
- [x] Rename `IEventQueue` and `PyPubSubEventQueue` with `Agent` to disambiguate (0d) - @shreyamalviya

### PR DESCRIPTION
**Pull request by shreyamalviya**
# What does this PR do?

Fixes a part of #2233 

## PR Checklist
* [ ] Have you added an explanation of what your changes do and why you'd like to include them?
* [ ] Is the TravisCI build passing?
* [ ] Was the CHANGELOG.md updated to reflect the changes?
* [ ] Was the documentation framework updated to reflect the changes?
* [ ] Have you checked that you haven't introduced any duplicate code?

## Testing Checklist

* [ ] Added relevant unit tests?
* [ ] Have you successfully tested your changes locally? Elaborate:
    > Tested by {Running the Monkey locally with relevant config/running Island/...}
* [ ] If applicable, add screenshots or log transcripts of the feature working

### CODE REVIEW

File: monkey/common/event_queue/types.py [hunk 1]
```diff
@@ -3,3 +3,4 @@
 from common.events import AbstractAgentEvent
 
 AgentEventSubscriber = Callable[[AbstractAgentEvent], None]
+IslandEventSubscriber = Callable[..., None]
```
Comment (mssalvatore) [inline monkey/common/event_queue/types.py:6]: Doesn't belong in common.

File: monkey/common/event_queue/i_island_event_queue.py [hunk 2]
```diff
@@ -1,12 +1,14 @@
...
 
-class IslandEventTopics(Enum):
-    AGENT_CONNECTED = auto()
-    CLEAR_SIMULATION_DATA = auto()
-    RESET_AGENT_CONFIGURATION = auto()
+
+class IslandEventTopic(Enum):
+    AGENT_CONNECTED = "agent_connected"
+    CLEAR_SIMULATION_DATA = "clear_simulation_data"
+    RESET_AGENT_CONFIGURATION = "reset_agent_configuration"
```
Comment (mssalvatore) [inline monkey/common/event_queue/i_island_event_queue.py:11]: These don't need values.

File: monkey/common/event_queue/pypubsub_island_event_queue.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    def publish(self, topic: IslandEventTopic, event_data: Any = None):
+        topic_value = topic.value  # needs to be a string for pypubsub
+
+        logger.debug(f"Publishing {topic_value} event")
+
+        # NOTE: `event_data` needs to match the MDS (message data specification) of the topic,
+        #       otherwise, errors will arise. The MDS of a topic is set when the topic is created,
+        #       which in our case is when a subscriber subscribes to a topic (in `subscribe()`)
+        #       which is new (hasn't been subscribed to before).
+        self._pypubsub_publisher.sendMessage(topicName=topic_value, event=event_data)
```
Comment (mssalvatore) [inline monkey/common/event_queue/pypubsub_island_event_queue.py:64]: We can look for opportunities to reduce duplication between this and the agent event queue. Maybe we could have a `PyPubSubPublisher` or similar that handles the details of pypubsub, and allow the event queues to manage topic->str conversion and other such things.

File: monkey/common/event_queue/pypubsub_island_event_queue.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+        #       provide an unsubscribe method (at this time) as subscriptions are expected to last
+        #       for the life of the process.
+        #
+        #       Specifically, if an instance object of a callable class is created and subscribed,
+        #       we don't want that subscription to end if the callable instance object goes out of
+        #       scope. Adding subscribers to self._refs prevents them from ever going out of scope.
+        self._refs.append(subscriber)
+
+    def publish(self, topic: IslandEventTopic, event_data: Any = None):
+        topic_value = topic.value  # needs to be a string for pypubsub
```
Comment (mssalvatore) [inline monkey/common/event_queue/pypubsub_island_event_queue.py:56]:
```suggestion
        topic_value = topic.name  # needs to be a string for pypubsub
```

**Review by mssalvatore — CHANGES_REQUESTED** [review]

**Review by ilija-lazoroski — APPROVED** [review]

**Review by mssalvatore — APPROVED** [review]

**Pull request merged**